coachmark/src/main/java/com/pseudoankit/coachmark/overlay/OverlayLayout.kt

```diff
@@ -0,0 +1,147 @@
+
+        // child count < 2 occurs on first and last coach mark
+        require(measurables.size <= 2) { "OverlayLayout cannot have more than two children" }
+
+        val gapTooltipScreenPx = gapTooltipScreen.roundToPx()
+
+        // measure children
+        val placeableCurrent = measure(configCurrent, OverlayChildLayoutId.CURRENT, measurables, constraints, gapTooltipScreenPx)
+        val placeablePrevious = measure(configPrevious, OverlayChildLayoutId.PREVIOUS, measurables, constraints, gapTooltipScreenPx)
+
+        // place children
+        layout(constraints.maxWidth, constraints.maxHeight) {
+            place(placeableCurrent, configCurrent)
+            place(placeablePrevious, configPrevious)
+        }
+    }
+}
+
+
+/**
+ * @return null if tooltipConfig is null
+ */
+private fun measure(
+    tooltipConfig: TooltipConfig?,
+    layoutId: OverlayChildLayoutId,
+    measurables: List<Measurable>,
+    constraintsParent: Constraints,
+    gapTooltipScreenPx: Int,
+): Placeable? {
+    if (tooltipConfig == null) return null
+
+    // constrain max width to prevent tooltip running off screen
+    var maxWidth = when (tooltipConfig.toolTipPlacement) {
+        ToolTipPlacement.Start -> {
+            constraintsParent.maxWidth - gapTooltipScreenPx - (constraintsParent.maxWidth - tooltipConfig.layout.startX.toInt())
+        }
+        ToolTipPlacement.End -> {
+            constraintsParent.maxWidth - gapTooltipScreenPx - tooltipConfig.layout.endX.toInt()
+        }
+
+        // Top and Bottom: allow full screen width, minus edge padding
+        else -> constraintsParent.maxWidth - (gapTooltipScreenPx shl 1)
+    }
+
+    /*
+        We can't currently constraint max height to prevent tooltip running off screen, because text
+        already uses as much width as it can by default, so there's no horizontal space to trade.
+    */
+
+    /*
+        This is mainly intended to avoid a crash when calculated max width is less than zero, such
+        as when tooltip is positioned to "End" while highlight is very close to right side of screen.
+        This fail-soft approach, where we display the tooltip running off the screen, allows the
+        developer to see the problem at runtime without having to check the log.
+     */
+    if (maxWidth < TOOLTIP_MAX_WIDTH_OVERRIDE_PX) {
+        maxWidth = Constraints.Infinity
+    }
+
+    val constraintsChild = Constraints(
+        minWidth = 0,
+        minHeight = 0,
+        maxWidth = maxWidth,
+        maxHeight = constraintsParent.maxHeight,
+    )
+
+    return measurables.find { it.layoutId == layoutId }?.measure(constraintsChild)
+}
+
+/**
+ * Centralizes null checks and switching on toolTipPlacement value.
+ * @param placeableOrNull no-op if null
+ * @param configOrNull no-op if null
+ */
+private fun Placeable.PlacementScope.place(placeableOrNull: Placeable?, configOrNull: TooltipConfig?) {
+    placeableOrNull?.let { placeable ->
+        configOrNull?.let { config ->
```

Comment:
this can be merged
if(placeableOrNull != null && configOrNull != null) to avoid two layer to nesting

Reply:
Good call.  It should smart cast in that case, which I didn't think of